### PR TITLE
feature(cli): add custom inspector proxy based on `metro-inspector-proxy`

### PR DESCRIPTION
**Pull request by byCedric**
# Why

Fixes ENG-7467

Related #21265

This is an initial draft to extend the CDP functionality of `metro-inspector-proxy`.

# How

The implementation is slightly wonky around the `ExpoInspectorDevice`. We want to reuse as much as possible from `metro-inspector-proxy`, but we need to add stateful data per device.

In order to achieve that, we generate a new class type, based on the user's installed `metro-inspector-proxy`. This makes everything less readable but should include future updates in these classes.

As for the `ExpoInspectorProxy`, to avoid having to do the same thing, we just wrap the whole inspector class and reuse what we can. The device map is "linked" within the original inspector proxy instance, making the data available to all methods that need it.

# Test Plan

Enable this feature with `EXPO_USE_CUSTOM_INSPECTOR_PROXY=1`

- [x] See tests for the actual CDP events we handle.
- [ ] See tests on the "bootstrapping code" to create the inspector and devices.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by Kudo — COMMENTED** [review]
leaving some nit comments. most of code looks good to me. thanks Cedric!

File: packages/@expo/cli/src/start/server/metro/resolveFromProject.ts [hunk 1]
```diff
@@ -45,6 +45,24 @@ export function importMetroResolverFromProject(
...
+  projectRoot: string
+): typeof import('metro-inspector-proxy') {
+  return importFromProject(projectRoot, 'metro-inspector-proxy');
+}
+
+/** Import `metro-inspector-proxy/src/Device` from the project. */
+export function importMetroInspectorDeviceFromProject(
+  projectRoot: string
+): typeof import('metro-inspector-proxy/src/Device') {
+  return importFromProject(projectRoot, 'metro-inspector-proxy/src/Device');
```
[the remark this answers is not in the document]
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/resolveFromProject.ts:59]: Unfortunately, `Device` is not exported from `metro-inspector-proxy`. So we do have to load it through `metro-inspector-proxy/src/Device`.

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/index.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          message: {
+            source: 'javascript',
+            level: 'info',
+            text: 'ooxx text',
+          },
+        },
+      };
+      socket?.send(JSON.stringify(payload));
+    });
+  }, 3000);
```
[the remark this answers is not in the document]
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/index.ts:44]: I did remove it, but had some issues with the inspector ack-ing that it's connected. So I might need to debug that just a bit more.

File: packages/@expo/cli/src/start/server/metro/resolveFromProject.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kudo) [reply packages/@expo/cli/src/start/server/metro/resolveFromProject.ts:59]: i was just thinking something like

```ts
function resolveMetroInspectorProxyFromProject(projectRoot: string) {
  const resolvedPath = resolveFrom.silent(projectRoot, 'metro-inspector-proxy');
  if (!resolvedPath) {
    throw new MetroImportError(projectRoot, 'metro-inspector-proxy');
  }
  return resolvedPath;
}

const metroInspectorProxyRoot = resolveMetroInspectorProxyFromProject('/path/to/app');
require(metroInspectorProxyRoot);
require(path.join(metroInspectorProxyRoot, 'src/Device'));
```

but that is not that useful. let's forget this comment 😂

**Comment by byCedric** [comment]
Hold off on merging until #21475 is merged

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        }
+
+        const device = this.devices.get(parseInt(deviceId, 10));
+        if (!device) {
+          throw new Error(`Device with ID "${deviceId}" not found.`);
+        }
+
+        debug('New debugger connected: device=%s, app=%s', device._name, device._app);
+
+        device.handleDebuggerConnection(socket, pageId);
```
Comment (byCedric) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:115]: Add a `socket.on('close')` to add a debug log whenever a debugger disconnects. E.g.

```suggestion
        device.handleDebuggerConnection(socket, pageId);
        socket.on('close', () => {
          debug('Debugger connected: device=%s, app=%s', device._name, device._app);
        });
```

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    return {
+      [WS_DEVICE_URL]: this.createDeviceWebSocketServer(),
+      [WS_DEBUGGER_URL]: this.createDebuggerWebSocketServer(),
+    };
+  }
+
+  private setServerAddress({ port, family, address }: AddressInfo) {
+    if (family === 'IPv6') {
+      this.metroProxy._serverAddressWithPort = `${address ?? '[::1]'}:${port}`;
```
Comment (Kudo) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:71]:
```suggestion
      this.metroProxy._serverAddressWithPort = `[${address ?? '::1'}]:${port}`;
```
the ipv6 address does include the brackets. my `server.address()` just returns `::1`.

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/device.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    /** Hook into the message life cycle to answer more complex CDP messages */
+    _interceptMessageFromDebugger(
+      request: DebuggerRequest<any>,
+      debuggerInfo: DebuggerInfo,
+      socket: WS
+    ) {
+      const result = this.onDebuggerMessage(request, debuggerInfo, socket);
+      return result
```
Comment (Kudo) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/device.ts:45]: i see a problem here that metro's `_interceptMessageFromDebugger` returns `boolean` only: https://github.com/facebook/metro/blob/4c520eda2ef322b1b8057d880a13ed776ecd3e79/packages/metro-inspector-proxy/src/Device.js#L457 and it doesn't send out the command returned in `result`. so from the devtools' protocol monitor, the `Network.getResponseBody` will not receive response through websocket.

**Review by Kudo — COMMENTED** [review]
leaving two problems from my testing. you could also test it on bare-expo with #21490.

be sure to set `USE_DEV_CLIENT = true` in **MainApplication** or `useDevClient = true` in **AppDelegate.mm**

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/device.ts (hunk 5 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/device.ts:45]: When adding vscode compatibility, I noticed this too. Will revise the types and docs to reflect better what methods should return! Thanks for flagging this

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts (hunk 4 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:71]: Oh man, gotta love IPv6. Thanks, will fix!

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/device.ts (hunk 5 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/device.ts]: Fixed, I also made the types a bit stricter around `Device`, cdp messages, and used `devtools-protocol` types in the network handler. Should get us better TS warnings related to protocol changes too :)

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/NetworkRespose.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import type { Protocol } from 'devtools-protocol';
+import { DebuggerInfo } from 'metro-inspector-proxy';
```
Comment (byCedric) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/NetworkRespose.ts:2]:
```suggestion
import type { DebuggerInfo } from 'metro-inspector-proxy';
```

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/__tests__/proxy.test.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      req: jest.fn() as any,
+      res: jest.fn() as any,
+      next: jest.fn(),
+    };
+
+    expoProxy.processRequest(ctx.req, ctx.res, ctx.next);
+    expect(spy).toBeCalledWith(ctx.req, ctx.res, ctx.next);
+  });
+
+  // TODO(cedric): add tests for `createWebSocketListeners`, using a mocked websocket
```
Comment (byCedric) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/__tests__/proxy.test.ts:24]: @EvanBacon We probably need something like https://www.npmjs.com/package/jest-websocket-mock for these tests to be effective.

Alternatively, I could split up the device initialization and test if devices are created successfully from URL. Unfortunately, that won't test the `.on('connection')` and `.on('close')` events.

Any thoughts on this one?

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/index.ts]: I just removed it, and it works fine without it. Probably related to: https://github.com/expo/expo/pull/21449#discussion_r1123633250

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/__tests__/proxy.test.ts (hunk 7 above)
Comment (EvanBacon) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/__tests__/proxy.test.ts:24]: the helper seems fine, I've never used it before.

File: packages/@expo/cli/src/start/server/metro/resolveFromProject.ts [hunk 8]
```diff
@@ -60,6 +55,25 @@ export function importMetroResolverFromProject(
   return importFromProject(projectRoot, 'metro-resolver');
 }
 
+/** Import `metro-inspector-proxy` from the project. */
+export function importMetroInspectorProxyFromProject(
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/resolveFromProject.ts:59]: Probably makes sense to generalize these functions going forward like: 

```ts
function importFrom<T>(projectRoot: string, id: string, assertion: string = ""): T {
 ...
}
```

File: packages/@expo/cli/src/utils/env.ts [hunk 9]
```diff
@@ -141,6 +141,11 @@ class Env {
   get EXPO_USE_PATH_ALIASES(): boolean {
     return boolish('EXPO_USE_PATH_ALIASES', false);
   }
+
+  /** **Experimental:** Use the network inspector by overriding the metro inspector proxy with a custom version */
+  get EXPO_USE_CUSTOM_INSPECTOR_PROXY(): boolean {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/utils/env.ts:146]: Add to the expo-cli docs https://docs.expo.dev/workflow/expo-cli/#environment-variables

File: packages/@expo/cli/src/start/server/metro/resolveFromProject.ts [hunk 10]
```diff
@@ -60,6 +55,25 @@ export function importMetroResolverFromProject(
...
+
+/** Import `metro-inspector-proxy/src/Device` from the project. */
+export function importMetroInspectorDeviceFromProject(
+  projectRoot: string
+): typeof import('metro-inspector-proxy/src/Device') {
+  return importFromProject(projectRoot, 'metro-inspector-proxy/src/Device');
+}
+
+/** Import `ws` from the project, installed by `metro-inspector-proxy` */
+export function importMetroInspectorWebSocketFromProject(projectRoot: string): typeof import('ws') {
```
[the remark this answers is not in the document]
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/resolveFromProject.ts:73]: This aligns the `ws` instance within the `device`. Right now, it uses the installed `ws` version from Metro in the project, to keep versions aligned. It's used here:
- https://github.com/expo/expo/pull/21449/files#diff-f9aca775ea37332e431e8f6d45b2cea39bc72931a91de437d9f2f78bfe556407R70
- https://github.com/expo/expo/pull/21449/files#diff-f9aca775ea37332e431e8f6d45b2cea39bc72931a91de437d9f2f78bfe556407R104

[the remark this answers is not in the document]
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/resolveFromProject.ts:73]: We could also say "screw it" and use a `ws` version from `@expo/cli` 🤷

File: packages/@expo/cli/src/utils/env.ts (hunk 9 above)
Comment (byCedric) [reply packages/@expo/cli/src/utils/env.ts:146]: Done

**Review by Kudo — APPROVED** [review]
tested this pr a little bit and it works like a charm. i don't have any further comments and leave the remaining code style/testability for Evan to review.

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+        debug('New debugger connected: device=%s, app=%s', device._name, device._app);
+
+        device.handleDebuggerConnection(socket, pageId);
+
+        socket.on('close', () => {
+          debug('Debugger disconnected: device=%s, app=%s', device._name, device._app);
+        });
+      } catch (error: unknown) {
+        console.log('ERRR', error);
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:126]: fixup

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        debug('Could not establish a connection to debugger:', error);
+        socket.close(WS_GENERIC_ERROR_STATUS, message || 'Unknown error');
+      }
+    });
+
+    return wss;
+  }
+}
+
+function asString(value?: string | string[]) {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:137]: add return type for conversion functions

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+      noServer: true,
+      perMessageDeflate: false,
+    });
+
+    // See: https://github.com/facebook/metro/blob/eeb211fdcfdcb9e7f8a51721bd0f48bc7d0d211f/packages/metro-inspector-proxy/src/InspectorProxy.js#L193
+    wss.on('connection', (socket, request) => {
+      try {
+        const { deviceId, pageId } = getExistingDeviceInfo(request.url);
+        if (!deviceId || !pageId) {
+          throw new Error(`Missing "device" and/or "page" IDs in query parameters`);
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:110]: Add a TODO to come back to these and change them to CommandError or related. If these happen often, then we should be handling them more gracefully.

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+// This is a workaround for `ConstructorType` not working on dynamically generated classes
+type Instantiatable<Instance> = new (...args: any) => Instance;
+
+export class ExpoInspectorProxy<D extends MetroDevice = MetroDevice> {
+  constructor(
+    public readonly metroProxy: MetroProxy,
+    private DeviceClass: Instantiatable<D>,
+    public readonly devices: Map<number, D> = new Map()
+  ) {
+    this.metroProxy = metroProxy;
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:22]:
```suggestion
```
I don't think this is required, the `public` part of the initializer should bind it to the class scope.

**Review by EvanBacon — APPROVED** [review]

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts (hunk 11 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:126]:
```suggestion
```

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts (hunk 12 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:137]:
```suggestion
function asString(value?: string | string[]): string {
```

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    wss.on('connection', (socket, request) => {
+      try {
+        const { deviceId, pageId } = getExistingDeviceInfo(request.url);
+        if (!deviceId || !pageId) {
+          throw new Error(`Missing "device" and/or "page" IDs in query parameters`);
+        }
+
+        const device = this.devices.get(parseInt(deviceId, 10));
+        if (!device) {
+          throw new Error(`Device with ID "${deviceId}" not found.`);
```
Comment (byCedric) [inline packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:115]:
```suggestion
          // TODO(cedric): change these errors to proper error types
          throw new Error(`Device with ID "${deviceId}" not found.`);
```

File: packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts (hunk 13 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts:110]:
```suggestion
          // TODO(cedric): change these errors to proper error types
          throw new Error(`Missing "device" and/or "page" IDs in query parameters`);
```

**Pull request merged**